src/clarity-angular/popover/popover.ts

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+/*
+ * Do NOT Angular this up. It assumes we're in the DOM, plays with native elements, ...
+ * It could potentially be used as part of clarity-ui as a vanilla Javascript helper.
+ *
+ * WARNING: This is a quick prototype version, use at your own risks.
+ */
+
+export enum Direction {
+    RIGHT
+}
+
+export interface PopoverOptions {
+    offsetX?: number;
+    offsetY?: number;
+    userAnchorParent?: boolean;
+}
+
+const POSITION_RELATIVE = "relative";
+const POSITION_ABSOLUTE = "absolute";
+const POSITION_FIXED = "fixed";
+
+const OVERFLOW_SCROLL = "scroll";
+const OVERFLOW_AUTO = "auto";
+
+interface InlineOverflow {
+    both: string;
+    x: string;
+    y: string;
+}
+
+const OVERFLOW_HIDDEN: InlineOverflow = {
+    both: "hidden",
+    x: "hidden",
+    y: "hidden"
+};
+
+export class Popover {
+    constructor(private element: any) {
+        // Browsers don't agree with what to do if some of these are not specified, so we set them all to be safe.
+        element.style.position = POSITION_ABSOLUTE;
+        element.style.top = 0;
+        element.style.left = 0;
+    }
+
+    public anchor(anchor: any, direction: Direction,
+                  {offsetX = 0, offsetY = 0, userAnchorParent = false}: PopoverOptions = {}) {
+        // TODO: we are assuming here that the popover is inside or next to the anchor.
+        // We'd need to go up the popover tree too otherwise
+        this.preventScrolling(anchor);
+        if (userAnchorParent) {
+            anchor = anchor.parentNode;
+        }
+        let anchorRect = anchor.getBoundingClientRect();
+        let popoverRect = this.element.getBoundingClientRect();
+        let leftDiff: number;
+        let topDiff: number;
+        // TODO: obviously, handle direction
+        switch (direction) {
+            case Direction.RIGHT:
+                leftDiff = anchorRect.left + anchorRect.width - popoverRect.left + offsetX;
```

Comment:
If you remove the `- popoverRect.left`, that means you're only positioning with respect to the anchor. In other words, you assume the popover's containing block (the closest parent with position absolute or relative) is the body. So it would break as soon as you have a positioned element anywhere up the tree.

Here is how I would explain what's happening here: by preventing parents* to scroll, we kind of "lock" the position of both the anchor and the popover on the screen. We have no idea what these positions are, because they depend on where the relative and absolute parents are, their own CSS styles, ... So the safest bet is simply to look at the **difference** of their positions, and adjust this with `transform: translate` to put the popover right next to the anchor. By using this difference, we do not rely in any way on parents or the rest of the DOM, the structure could be super complex and it would still position correctly.


* It's a little more involved than just all parents, see the walk-up-the-tree from `preventScrolling()` for more info.